design-proposals/pvDiskStorage.md

```diff
@@ -0,0 +1,132 @@
+
+
+## API
+
+This section is concerned about how the Persistent Volumes are referenced
+in the `VM` Resource type.
+
+In general the referencing is aligned with how pods are consuming Persistent
+Volume Claims as described [here](https://kubernetes.io/docs/api-reference/v1.5/#persistentvolumeclaimvolumesource-v1)
+
+Today the `VM.spec.domain` reflects much of
+[libvirt's domain xml specification](http://libvirt.org/formatdomain.html#elementsDisks).
+To support Persistent Volumes as a backing store for file storage, two changea are
+proposed:
+
+1. an additional disk type `PersistentVolumeClaim` is accepted
+   If this type is used, then
+2. an additional `claimName` attribute needs to be specified
+   which is specifying the claim to use
+
+Example with the following PV and PVC:
+
+```yaml
+kind: PersistentVolume
+metadata:
+  name: pv001
+  labels:
+    release: "stable"
+spec:
+  capacity:
+    storage: 5Gi
+  iscsi:
+    targetPortal: example.com:3260
+    iqn: iqn.2013-07.com.example:iscsi-nopool/
+    lun: 0
+---
+kind: PersistentVolumeClaim
+metadata:
+  name: disk-01
+spec:
+  resource:
+    requests:
+      storage: 4Gi
+  selector:
+    matchLabels:
+release: "stable"
+```
+
+this is used by the Vm in the following way:
+
+```yaml
+kind: VM
+spec:
+  domain:
+    devices:
+      disks:
+      - type: PersistentVolumeClaim
+        claimName: disk-01
+        bus: scsi
+        target: sda
+```
+
+
+## Implementation
+
+### Flow
+
+1. User adds the PersistentVolumeClaim as described above to the VM
+   instance.
+2. The VM Pod is getting scheduled on a host, the `virt-handler`
+   identifies the Claim, translates it into the corresponding
+   libvirt representation and includes it in the domain xml.
+
+Note: The `virt-controller` does not do anything with the claim.
+
+In this flow, the claim is not used by the pod, the claim is only used
+by the `virt-handler` to identify the conneciton details to the storage.
+
+Because VMs only accept block storage as disks, the handler can only
+accept claims which are backed by block storage types.
+
+
+### `virt-handler` changes
+
+Once a VM is scheduled on a host, the `virt-handler` is transforming the
+VM Spec into a libvirt domain xml.
+
+During this transformation, every disk which is of type `PersistentVolumeClaim`
+needs to be transformed into an adequate libvirt disk type.
+
+When the VM Pod got started, it got the named volume claim attached. The
+responsibility of the `virt-handler` is now to make it available to `qemu`.
+
+To achieve this, the `virt-handler` needs to look at volume attached to a
+claim, to identify the `type` and connection deails.
+If the type is either iSCSI or RBD, then it can take the connection details
+and transform them into the correct libvirtd representation.
+
+Example for an iSCSI volume:
+
+```xml
+  <disk type='network' device='disk'>
+    <driver name='qemu' type='raw'/>
+    <source protocol='iscsi' name='iqn.2013-07.com.example:iscsi-nopool/2'>
```

Comment:
The trick here is, that libvirt will use qemu's facilities to connect to the iscsi target. Thus the host is not involved, and to the VM it will look like a regular disk.

And both things are intended: Exclude the host - this helps to avoid all kind of udev issues we have with containers - and second make it transparent to the vm how the storage is implemnted.